src/metabase/lib/query.cljc

```diff
@@ -64,6 +64,21 @@
   (and (mc/validate ::lib.schema/query query)
        (boolean (can-run-method query))))
 
+(defmulti can-save-method
+  "Returns whether the query can be saved based on first stage :lib/type."
+  (fn [query]
+    (:lib/type (lib.util/query-stage query 0))))
+
+(defmethod can-save-method :default
+  [_query]
+  true)
+
+(mu/defn can-save :- :boolean
+  "Returns whether the query can be saved."
+  [query :- ::lib.schema/query]
```

Comment:
I don't understand this part exactly.
For a non-native query, will it _always_ return true? I am not sure this is how it has to be but this is how I implemented it in a JS version 😁

Reply:
for mbql queries it will be the same as `can-run`. Is it correct?

Reply:
Given that it didn't even exist before, I don't think so. Perhaps for MBQL it should always be `true`, this was my assumption.

Reply:
Discussed - it should be `can-run` which checks whether the query is not empty in this case.

Reply:
I see that you've discussed changing the default value; that makes sense to me.

For understanding, one way to think about how `defmulti` + `defmethod` work is that it resembles a `switch`/`case` with its code scattered across files. The function on the `defmulti` is the `switch (...)` argument, it returns the value we're switching on. Then you specify `case`s for particular values with `(defmethod lib.query/can-save-method :mbql.stage/native ...)`, as we do above in `metabase.lib.native`.

If none of the overrides matches, the `:default` case is used instead, just like the `default:` clause of a `switch` statement.

(There's more to multimethods, but that's a solid analogy for basic uses like this one.)